File: Solution257.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> results = new ArrayList<String>();
        helper(root, new ArrayList<TreeNode>(), results);
        return results;
    }

    private void helper(TreeNode node, List<TreeNode> list, List<String> results) {
        if (node == null) return;

        list.add(node);
        if (node != null && node.left == null && node.right == null) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < list.size(); i++) {
                sb.append(list.get(i).val);
                if (i != list.size() - 1) sb.append("->");
            }
            results.add(sb.toString());
        }
        helper(node.left, list, results);
        helper(node.right, list, results);
        list.remove(list.size() - 1);
        return;
    }
}

File: rust/src/solution257.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}
use crate::common::Solution;
use std::rc::Rc;
use std::cell::RefCell;
use std::vec;
impl Solution {
    pub fn binary_tree_paths(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<String> {
        let mut results = vec![];
        Self::helper(root, &mut results, &mut vec![]);
        results
    }

    fn helper(node: Option<Rc<RefCell<TreeNode>>>, results: &mut Vec<String>, result: &mut Vec<String>){
        match node {
            None => (),
            Some(n) => {
                let current = Rc::try_unwrap(n).unwrap().into_inner();

                if current.left.is_none() && current.right.is_none() {
                    result.push(current.val.to_string());
                    results.push(result.join(""));
                } else {
                    result.push(current.val.to_string() + "->");
                }

                Self::helper(current.left, results, result);
                Self::helper(current.right, results, result);
                result.remove(result.len() - 1);
            }
        }
    }
}